Share token generate- Separate branch

nikitadhiman opened the pull request:
Pull Request is for creation of a share token and storing that token in the #__share_draft database.

Summary of Changes:
1. A new button is created for 'share', this button would be visible only when article is saved and is placed next to version button.
2. A function shareTokenGenerate() and shareDraft() are created in the controller, shareTokenGenerate() will generate a random token containing a-z,A-Z and 0-9 but no special characters. 
3. In model file, query has been made to store this token in the database.

Testing Instruction:
1. Apply the patch, make sure that the #__share_draft db exists. Have used JTables for the same
2. create a new article and save the article.
3.Once saved, hit share button, the values that should be stored the db are articleId, title of the article and newly generated random token.

Test 1 results:
1. When the table exists:
{"success":true,"message":"Token Saved","messages":null,"data":"IeOrhUISqtAm3PXV"}
2. When table does not exists:
{"success":false,"message":"Token not Saved","messages":null,"data":false}

Tasks:
- [x] Create controller file article.json.php
- [x] Fix the $db in the article model 
- [x] Fix the storing of token link to store created date and time
- [x] Remove the storing of title
- [x] Remove the button at the bottom of the article edit page
- [x] Move the div below the article title
- [x] Alpha-order language strings
- [x] Fix the share button URL
- [x] Return share link from article model
- [x] Check existing token based on article ID
- [x] Make link clickable
- [x] Complete step 2 from the mentioned steps[review needed]

nikitadhiman committed:
share token generation

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update en-GB.com_content.ini

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update en-GB.com_content.ini

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update view.html.php

nikitadhiman committed:
share using JTable

nikitadhiman committed:
proper alignment in tables/share.php

nikitadhiman committed:
proper alignment in tables/share.php

nikitadhiman committed:
proper alignment in tables/share.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
shared articles manager-Part 1

nikitadhiman committed:
Merge remote-tracking branch 'origin/staging' into staging

nikitadhiman committed:
shared articles manager-Part 1

nikitadhiman committed:
Update article.php

nikitadhiman committed:
updated the model part

nikitadhiman committed:
Merge remote-tracking branch 'origin/staging' into staging

nikitadhiman committed:
travis issue controller

nikitadhiman committed:
separate branch for token generation

nikitadhiman committed:
removed unnecessary changes

nikitadhiman commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,31 @@ public function batch($model = null)
+
+		if ($return)
+		{
+			JText::_('COM_CONTENT_TOKEN_SAVED');
```
@roland-d , the ajax call would be from controller and would include what all we discusses yesterday,right?

roland-d replied:
@nikitadhiman The AJAX call will be initiated from the view by clicking on a button.

nikitadhiman replied:
@roland-d , does this have the same 8 odd steps you mentioned once; passing the data and all

roland-d replied:
@nikitadhiman You will have to return a JResponseJson object here.

nikitadhiman committed:
js file, share-uncompressed

nikitadhiman committed:
button

nikitadhiman committed:
share.js file, needs some corrections

nikitadhiman committed:
updated

nikitadhiman committed:
Merge branch 'staging' of https://github.com/joomla-projects/gsoc16_shareable-draft-content into ajax_call_button

nikitadhiman committed:
checking security token

nikitadhiman committed:
moved $title

roland-d committed:
Clean up file

nikitadhiman committed:
edit.php changes

nikitadhiman committed:
Merge remote-tracking branch 'origin/ajax_call_button' into ajax_call_button

Conflicts:
	administrator/components/com_content/views/article/tmpl/edit.php

nikitadhiman committed:
edit.php changes

nikitadhiman committed:
few js changes

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Merge branch 'ajax_call_button' into share_token_generate

nikitadhiman committed:
minor

nikitadhiman committed:
Merge remote-tracking branch 'origin/share_token_generate' into share_token_generate

yvesh commented on media/system/js/share-uncompressed.js:
```diff
@@ -0,0 +1,39 @@
+    {
+        var share_draft = {
+            articleId  : $('#jform_id').val(),
+            title: $('#jform_title').val(),
```
Btw.: When you want to use $, you need to make sure jQuery is used with $ (There could be also be mootools), either with $ = jQuery or surrounding your whole document ready function with:

(function($){
$(document).ready...
...
// End of your code
})(jQuery);

I would suggest using the last one here.

yvesh commented on media/system/js/share-uncompressed.js:
```diff
@@ -0,0 +1,39 @@
+                    window.scrollTo(0, 0);
+                }
+
+    });
```
You should also add an error fallback here. Else the user will never notice something went wrong and don't get any reaction to clicking the button. Just add (concenate) it after done

``` javascript
.done .. {(
// ...
}).fail(function(xhr, ajaxOptions, thrownError) {
    // Just illustration, you should really improve that output etc, add error type
    Joomla.renderMessages(xhr.responseText);
}); 
```

yvesh commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,40 @@ public function batch($model = null)
+		else
+		{
+			JText::_('COM_CONTENT_TOKEN_ERROR');
+		}
```
This if / else is still useless :-)

You need to output your shared token here, maybe with a status code. So this is your target json:

`{"status": "ok", "token": "aasdfasdfsadf"}`

To achieve that you have to update your `$model->shareToken` method. For the status you can just check if the token is not null / empty etc for the beginning.

```php
$result = array(
    'status' => $status,
    'token' => $token
);
```

`echo new JResponseJson($return);` // Or json_encode($return);

roland-d replied:
@nikitadhiman Can you remove the if/else here?

yvesh commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,53 @@ public function featured($pks, $value = 0)
+		catch (Exception $e)
+		{
+			throw new RuntimeException($e->getMessage());
+		}
```
Imo i would remove the try catch block here, or return null / '' empty string if somethign fails (After logging the error)

yvesh commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,53 @@ public function featured($pks, $value = 0)
+			$db->setQuery($query);
+			$table->save($data);
+
+			return $db->execute();
```
Imo don't return the boolean here.

I would return the token, e.g. return $token;

nikitadhiman replied:
yeah right, will change this

yvesh commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,53 @@ public function featured($pks, $value = 0)
+	 *
+	 * @param   string  $title  The title of the shared article.
+	 *
+	 * @return  boolean  True on success.
```
Imo the method should return the generated token (as you are going to need that after the ajax call) and not the boolean.

/cc @roland-d you know more about the planed implementation. Can you comment?

yvesh commented on administrator/components/com_content/tables/share.php:
```diff
@@ -0,0 +1,30 @@
+	 */
+	public function __construct(&$db)
+	{
+		parent::__construct('#__share_draft', 'id', $db);
```
As the table is related to com_content (imo) i would suggest you use #__content as prefix.

e.g. call it #__content_share or something like that (would also match you class name)

nikitadhiman replied:
this means changing the name from #__share_draft to #__content_share?

yvesh replied:
I think that name would be more fitting, but you should ask your mentors on that :)

nikitadhiman committed:
controller-model changes

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,30 @@ public function batch($model = null)
+		if (!JSession::checkToken('get'))
+		{
+			$this->app->enqueueMessage(JText::_('JINVALID_TOKEN'));
+			$this->app->redirect('index.php');
```
You are inside an AJAX call, no need to do a redirect. You need this code:

```
$app->enqueueMessage(JText::_('JINVALID_TOKEN'), 'error');
            echo new JResponseJson;
            $app->close();
```

You also need to declare the $app otherwise you can't use it. $this->app doesn't exist.
You

nikitadhiman replied:
done this, please check

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,30 @@ public function batch($model = null)
+			$this->app->redirect('index.php');
+		}
+
+		$title = $this->input->getArray(array('jform' => array('title' => 'string')));
```
This code from $title until $return should be wrapped in a try/catch so we can catch any errors generated in the model.

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,30 @@ public function batch($model = null)
+		// Get the model
+		$model = $this->getModel();
+		$return = $model->shareToken($title);
+		echo new JResponseJson($return);
```
You need to add the parameter if there was an error in the JResponseJson.

nikitadhiman replied:
for eg like this ?
echo new JResponseJson($return, JText::_('MY_TASK_ERROR'), true); 
the last parameter is to manually set the error flag, should the last parameter be there too?

roland-d replied:
@nikitadhiman `echo new JResponseJson($return, JText::_('MY_TASK_ERROR'), true);` indeed like that. Only the echo needs to come after the try/catch. Just set the variables here and use the variables in the echo statement.

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update share-uncompressed.js

nikitadhiman committed:
Update share-uncompressed.js

nikitadhiman commented on media/system/js/share-uncompressed.js:
```diff
@@ -0,0 +1,40 @@
+    });
+});
+}(jQuery);
+
```
@roland-d , please review this.
I guess $document requires declaration too 
$document = JFactory::getDocument();
$document->addScriptDeclaration('

cc: @yvesh

yvesh replied:
@nikitadhiman This is a JavaScript file, so you don't need that PHP code here. You need to load it (from your view) with

`JHtml::_('script', 'path/to/file' ...)`

See Rolands sample here: 

https://github.com/joomla-extensions/boilerplate/commit/64d594472412086a1a6a7a492ff45e53a5538a00

// Edit: You actually do that already [here](https://github.com/joomla-projects/gsoc16_shareable-draft-content/pull/13/files#diff-68c7d014db78a0a04120699ace86a9bfR86)

nikitadhiman replied:
thanks @yvesh :)

nikitadhiman committed:
Update article.php

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,39 @@ public function batch($model = null)
+		{
+		echo new JResponseJson($e);
+		}
+	}
```
The echo would be here:

```
echo new JResponseJson($response, $message, $error)
$app->close;
```

nikitadhiman replied:
one doubt, $message= JText::_('MY_TASK_ERROR') and this would be called in echo new JResponseJson($reponse,$message,$error) and from where would the $error be fetched?

roland-d replied:
You set a default value before the try catch and only set a new value in the catch. Same with $message.

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,39 @@ public function batch($model = null)
+	{
+		if (!JSession::checkToken('get'))
+		{
+			$app    = JFactory::getApplication();
```
The $app needs to moved above the if statement, so it can be used later.

nikitadhiman committed:
Update share-uncompressed.js

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update article.php

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,47 @@ public function batch($model = null)
+		{
+		$error = true;
+		$message = JText::_('COM_CONTENT_TOKEN_ERROR');
+		echo new JResponseJson($e);
```
This doesn't belong here because you output the resul on line 185.

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,47 @@ public function batch($model = null)
+		// Get the model
+		$model = $this->getModel();
+		$return = $model->shareToken($title);
+		JFactory::getApplication()->close();
```
This has to be removed because you are closing the application on line 186.

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,47 @@ public function batch($model = null)
+		echo new JResponseJson($e);
+		}
+		
+		echo new JResponseJson($return, $message, $error);
```
The $return may not exist if an error is caught, so you need to set a default for this as well.

nikitadhiman replied:
what can be the default value of $return?

roland-d replied:
false, because if all is good the return will be set to true. The action is not successful unless told it is successful.

roland-d commented on media/system/js/share.js:
```diff
@@ -0,0 +1,31 @@
+jQuery(document).ready(function(e){e("#sharebutton").click(function()
```
This file must be compressed. You can use jscompress.com to compress the content.

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update share.js

nikitadhiman committed:
Update share-uncompressed.js

nikitadhiman committed:
Update article.php

nikitadhiman committed:
Update edit.php

nikitadhiman committed:
Update article.php

nikitadhiman commented on administrator/components/com_content/views/article/tmpl/edit.php:
```diff
@@ -82,6 +82,11 @@
+JHtml::script('system/share.js', false, true);
+JFactory::getDocument()->addScriptDeclaration('
+ 	var sharebutton_url = "' . addslashes(JUri::base()) . 'index.php?option=com_content&task=article.shareDraft&format=json&' . JSession::getFormToken() . '=1";
+  ');
```
@roland-d @yvesh , Unable to store the values in table despite of table being present. What do you suggest?

nikitadhiman committed:
minor

nikitadhiman committed:
Update edit.php

nikitadhiman committed:
Update edit.php

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,46 @@ public function batch($model = null)
+		
+		if (!JSession::checkToken())
+		{
+			$app->enqueueMessage(JText::_('JINVALID_TOKEN'));
```
This response needs to be set in the new JResponseJson instantiation.

roland-d commented on administrator/components/com_content/controllers/article.php:
```diff
@@ -145,6 +146,46 @@ public function batch($model = null)
+		if (!JSession::checkToken())
+		{
+			$app->enqueueMessage(JText::_('JINVALID_TOKEN'));
+			echo new JResponseJson;
```
The response needs an error status set. Now it has no error status as if everything is OK.

nikitadhiman committed:
validation to check if id is already present

nikitadhiman committed:
travis fixes

nikitadhiman committed:
minor

nikitadhiman committed:
minor

nikitadhiman committed:
mino

nikitadhiman committed:
controller update

yvesh commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,70 @@ public function featured($pks, $value = 0)
+           		$token = $this->shareTokenGenerate();
+           		$data = array('articleId' => $this->get('id'), 'title' => $title, 'sharetoken' => $token);
+           		$table->save($data);
+           		return $token;
```
New line before return

yvesh commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,70 @@ public function featured($pks, $value = 0)
+           		$db->execute();
+           		$valueshow = $db->loadObject();
+           		echo "not a new value, the existing value is:" . $valueshow;
+       		}
```
Indention

yvesh commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,70 @@ public function featured($pks, $value = 0)
+
+           		$db->execute();
+           		$valueshow = $db->loadObject();
+           		echo "not a new value, the existing value is:" . $valueshow;
```
you need to return the existing value here. return $valueshow; (which is a pretty bad variable name)

yvesh commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,70 @@ public function featured($pks, $value = 0)
+             		->from($db->quoteName('#__share_draft'))
+             		->where($db->quoteName('articleId') . '=' . $db->quote($this->get('id')));
+
+       		$db->execute();
```
Indentation

nikitadhiman committed:
article.json.php file

nikitadhiman committed:
moving div tag

nikitadhiman committed:
minor

nikitadhiman committed:
minor update

nikitadhiman committed:
removed button

nikitadhiman committed:
minor

Ruchiranga commented on media/system/js/share-uncompressed.js:
```diff
@@ -0,0 +1,46 @@
+
+                        window.scrollTo(0, 0);
+                    }
+                }
```
You are missing a closing bracket here and there is an additional closing bracket in line 42. :)
The ajax call should be in the format
$.ajax().done(function(){}).fail(function(){});

nikitadhiman committed:
updated file

nikitadhiman committed:
compressed file

nikitadhiman committed:
Storing date and time in #__share_draft

Now the date and time value is stored and title of the article is not stored in db

nikitadhiman committed:
not storing title value

nikitadhiman committed:
language file

nikitadhiman committed:
alpha order language string

nikitadhiman committed:
comment correction

nikitadhiman committed:
moving the shareDraft to json.php file

nikitadhiman committed:
moved the content to json controller file

nikitadhiman committed:
minor

nikitadhiman committed:
minor

roland-d commented on administrator/components/com_content/controllers/article.json.php:
```diff
@@ -0,0 +1,56 @@
+		
+	    	      if (!JSession::checkToken())
+	    	      {
+	      		        echo new JResponseJson(JText::_('JINVALID_TOKEN'));
```
This construct is wrong because the first parameter needs to be set to null because there is no dta to be returned, second parameter is the message and the third parameter needs to be set to true or false to indicate if an error occurred

nikitadhiman replied:
ok will update this

nikitadhiman committed:
JResponseJson update

nikitadhiman committed:
minor

roland-d commented on administrator/components/com_content/controllers/article.json.php:
```diff
@@ -0,0 +1,56 @@
+		
+	    	      if (!JSession::checkToken())
+	    	      {
+	      		        echo new JResponseJson(null, JText::_('JINVALID_TOKEN'), false);
```
Are you sure that no error has occurred?

nikitadhiman committed:
fixing $db in model file-review needed

nikitadhiman committed:
minor

nikitadhiman committed:
This solves the model bug returning $token

nikitadhiman committed:
js update in edit.php

nikitadhiman committed:
js fixing...testing

nikitadhiman committed:
js fixing...testing

nikitadhiman committed:
js fixing...testing

nikitadhiman committed:
for testing purpose

nikitadhiman committed:
minor

nikitadhiman committed:
solving invalid token error

nikitadhiman committed:
minor

nikitadhiman committed:
this solves all token related issues

nikitadhiman committed:
travis fix

nikitadhiman committed:
minor

nikitadhiman committed:
dealing with repeated token error

nikitadhiman committed:
minor

nikitadhiman committed:
update to solve repeated token issue

nikitadhiman committed:
storing date of token creation

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,59 @@ public function featured($pks, $value = 0)
+	public function shareToken()
+	{
+		$table = $this->getTable('Share', 'ContentTable');
+		$jinput = JFactory::getApplication()->input;
```
This belongs in the controller

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,59 @@ public function featured($pks, $value = 0)
+		$jinput = JFactory::getApplication()->input;
+		$articleId = $jinput->get('id', 0);
+		
+        	$db = $this->getDbo();
```
This variable is already available in the models via $this->_db

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,59 @@ public function featured($pks, $value = 0)
+        	$query
+             		->select('sharetoken')
+             		->from($db->quoteName('#__share_draft'))
+             		->where($db->quoteName('articleId') . '=' . $db->quote($articleId));
```
This is an integer, so we should treat it like such using (int) instead of $db->quote

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,59 @@ public function featured($pks, $value = 0)
+        	$query = $db->getQuery(true);
+
+        	$query
+             		->select('sharetoken')
```
Missing quoteName() here

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,59 @@ public function featured($pks, $value = 0)
+	 */
+	public function shareToken()
+	{
+		$table = $this->getTable('Share', 'ContentTable');
```
This should only be loaded if there is no token.

nikitadhiman replied:
To check existing values, still can be done without this?

roland-d replied:
Yes, because you don't use it.

roland-d commented on media/system/js/share-uncompressed.js:
```diff
@@ -0,0 +1,42 @@
+ */
+var shareDraft;
+
+jQuery(document).ready(function($)
```
This is not needed, we can just make a plain ```function shareDraft()``` function

roland-d commented on media/system/js/share-uncompressed.js:
```diff
@@ -0,0 +1,42 @@
+	shareDraft = function () {
+		var share_draft = {
+			id: articleId,
+			title: $('#jform_title').val()
```
Obsolete since we don't use it anymore

nikitadhiman replied:
yes, will remove this

roland-d commented on administrator/components/com_content/views/article/tmpl/edit.php:
```diff
@@ -82,12 +88,20 @@
 $layout = $isModal ? 'modal' : 'edit';
 $tmpl = $isModal ? '&tmpl=component' : '';
+
+JFactory::getDocument()->addScriptDeclaration('
```
You can create the full URL here instead of fragments here and then construct it in the shareDraft() function.

roland-d commented on administrator/components/com_content/views/article/tmpl/edit.php:
```diff
@@ -82,12 +88,20 @@
+	var sessionToken = "' . JSession::getFormToken() . '";
+	var articleId    = "' . $this->item->id . '";
+  ');
+JHtml::script('system/share-uncompressed.js', false, true);
```
This needs to be share.js, Joomla will add uncompressed if needed.

nikitadhiman replied:
this is for testing purpose, will remove it after compressed file is updated

roland-d replied:
You need to enable the Joomla debug mode and the uncompressed version will be loaded.

roland-d commented on administrator/components/com_content/views/article/tmpl/edit.php:
```diff
@@ -109,15 +123,15 @@
 			<div class="row-fluid form-horizontal-desktop">
 				<div class="span6">
-					<?php echo JLayoutHelper::render('joomla.edit.publishingdata', $this); ?>
+					<?php echo JLayoutHelper::render('joomla.edit.publishingta', $this); ?>
```
Why is this changed?

nikitadhiman replied:
not my changes :|

roland-d replied:
It is your pull request :)

roland-d replied:
Still need to fix this typo

roland-d commented on administrator/components/com_content/tables/share.php:
```diff
@@ -0,0 +1,30 @@
+	 *
+	 * @param   JDatabaseDriver  &$db  Database connector object
+	 *
+	 * @since   1.6
```
This should be __DEPLOY_VERSION__ instead of 1.6

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,59 @@ public function featured($pks, $value = 0)
+	 *
+	 * @return  token.
+	 *
+	 * @since   3.7
```
This should be `__DEPLOY_VERSION__` instead of 3.7

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,59 @@ public function featured($pks, $value = 0)
+	 *
+	 * @return  boolean  True on success.
+	 *
+	 * @since 3.7
```
This should be `__DEPLOY_VERSION__` instead of 3.7

nikitadhiman committed:
changing @since in model

nikitadhiman committed:
this returns the complete link

nikitadhiman committed:
removing unnecessary data

nikitadhiman committed:
The link is clickable now!

nikitadhiman committed:
updated compressed file

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,61 @@ public function featured($pks, $value = 0)
+      			
+      		 }
+      		 
+      		$link = JUri::root() . 'index.php?option=com_content&view=article&id=' . $articleId . '&share=' . $token;
```
Why are you not using the Joomla API here as discussed? JHtml::_('link', ...

roland-d replied:
Don't forget this.

nikitadhiman committed:
makes link magical

nikitadhiman commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,61 @@ public function featured($pks, $value = 0)
+      			
+      		 }
+      		 
+      		$link = JUri::root() . 'index.php?option=com_content&view=article&id=' . $articleId . '&share=' . $token;
```
I tried.....it didn't worked :P or I did something wrong :|

roland-d replied:
Without specifying what you did, I can't give you an answer.

nikitadhiman committed:
changed share-uncompressed to share

roland-d commented on components/com_content/models/article.php:
```diff
@@ -148,8 +148,21 @@ public function getItem($pk = null)
+        			$queryy
+             				->select($db->quoteName('sharetoken'))
+             				->from($db->quoteName('#__share_draft'))
+             				->where($db->quoteName('articleId') . '=' . $pk);
```
We should check on the token value rather than the ID, so you don't have to keep comparing the value.

nikitadhiman replied:
the token value comparison is done in next lines....this is to fetch the sharetoken for present articleId

roland-d replied:
That is what I am saying, you don't need to do the token check if you use the token in the where clause.

roland-d commented on components/com_content/models/article.php:
```diff
@@ -164,7 +177,7 @@ public function getItem($pk = null)
 
 				// Check for published state if filter set.
-				if (((is_numeric($published)) || (is_numeric($archived))) && (($data->state != $published) && ($data->state != $archived)))
+				if ( !($haveToken == $currentToken) && ((is_numeric($published)) || (is_numeric($archived))) && (($data->state != $published) && ($data->state != $archived)))
```
This is not completely correct, because the article would still load if it is archived. We should only check the token if the article is unpublished.

nikitadhiman replied:
this means that token check would be performed only if status=0 else no need for token check[as article is already published].....and then rest of the code?

roland-d replied:
Correct. The rest you don't need to do anything else.

nikitadhiman committed:
Article= archieved but shared, 404 error is set

This fulfills all the conditions of the draft being published...Tested successfully!

nikitadhiman committed:
minor

roland-d commented on components/com_content/models/article.php:
```diff
@@ -177,9 +177,13 @@ public function getItem($pk = null)
+				if (((is_numeric($published)) || (is_numeric($archived))) && (($data->state != $published) && ($data->state != $archived)))
 				{
-					return JError::raiseError(404, JText::_('COM_CONTENT_ERROR_ARTICLE_NOT_FOUND'));
+					if( !($haveToken == $currentToken))
```
This should be part of the if statement. As mentioned earlier, using the token in the database query, you don't have to keep doing the comparison. It then just becomes a true/false value.

nikitadhiman committed:
moving jinput to controller

nikitadhiman committed:
now, model knows nothing about outside world

nikitadhiman committed:
$this->_db in model

nikitadhiman committed:
@param in comment

nikitadhiman committed:
uncompressed file

nikitadhiman committed:
comparing token instead in where clause

nikitadhiman committed:
shareDraft() in compressed file

nikitadhiman committed:
EOF in uncompressed

nikitadhiman committed:
minor

nikitadhiman committed:
new compressed file

roland-d commented on components/com_content/models/article.php:
```diff
@@ -166,7 +179,11 @@ public function getItem($pk = null)
 				if (((is_numeric($published)) || (is_numeric($archived))) && (($data->state != $published) && ($data->state != $archived)))
 				{
-					return JError::raiseError(404, JText::_('COM_CONTENT_ERROR_ARTICLE_NOT_FOUND'));
+					if ($haveToken == 0)
```
This needs to be merged with the parent IF statement

roland-d commented on administrator/components/com_content/controllers/article.json.php:
```diff
@@ -0,0 +1,58 @@
+		 {
+				  $app    = JFactory::getApplication();
+				  $jinput = JFactory::getApplication()->input;
+        			  $articleId = $jinput->get('id', 0);
```
Filter this on int, so you are sure the value is an integer.

roland-d commented on administrator/components/com_content/controllers/article.json.php:
```diff
@@ -0,0 +1,58 @@
+				  }
+				  catch (Exception $e)
+				  {
+						  $error = true;
```
These lines are indented too far.

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,60 @@ public function featured($pks, $value = 0)
+	 *
+	 * @since  _DEPLOY_VERSION_
+	 */
+	public function shareToken($articleId)
```
What happens if the articleId is 0?

nikitadhiman replied:
we display message saying it's a new article?

roland-d replied:
No, you should return a message saying that the article ID is invalid

nikitadhiman replied:
right, will include that!

nikitadhiman replied:
one doubt, will that be done using try-catch or simple if-else?

roland-d replied:
You will do that with an if statement, no else is needed here.

nikitadhiman replied:
understood!

nikitadhiman replied:
Is this correct?
have added the articleId check

nikitadhiman committed:
typo

nikitadhiman committed:
minor

nikitadhiman committed:
publishing draft

nikitadhiman committed:
final changes

roland-d commented on media/system/js/share-uncompressed.js:
```diff
@@ -0,0 +1,41 @@
+			}
+		})
+		.fail(function (xhr, ajaxOptions, thrownError) {
+			alert("fail"),
```
Why is this alert still here?

nikitadhiman replied:
was testing!

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -706,6 +705,60 @@ public function featured($pks, $value = 0)
+             		->from($this->_db->quoteName('#__share_draft'))
+             		->where($this->_db->quoteName('articleId') . '=' . $this->_db->quote($articleId));
+
+       		$db->setQuery($query);
```
This doesn't work, $db is undefined here.

nikitadhiman committed:
proper indent

nikitadhiman committed:
minor

nikitadhiman committed:
removed alert from js

nikitadhiman committed:
minor

nikitadhiman committed:
if articleId is 0

nikitadhiman committed:
added language string

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -730,6 +730,13 @@ public function shareTokenGenerate()
+		//checking for valid articleId
+		if ($articleId == 0)
+		{
+			$this->setError(JText::_('COM_CONTENT_INVALID_ARTICLEID'));
```
@nikitadhiman setError() is deprecated, why do you use it? You need to throw an exception here.

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -730,6 +730,13 @@ public function shareTokenGenerate()
+			$this->setError(JText::_('COM_CONTENT_INVALID_ARTICLEID'));
+			
+			return false;
+		}
```
Codestyle, you need an empty line after this.

nikitadhiman replied:
ok will add

roland-d commented on administrator/language/en-GB/en-GB.com_content.ini:
```diff
@@ -112,6 +112,7 @@ COM_CONTENT_FLOAT_LABEL="Image Float"
 COM_CONTENT_FLOAT_INTRO_LABEL="Intro Image Float"
 COM_CONTENT_HEADING_ASSOCIATION="Association"
 COM_CONTENT_ID_LABEL="ID"
+COM_CONTENT_INVALID_ARTICLEID="Invalid ArticleId"
```
Change the text to "An invalid article ID has been found"

nikitadhiman replied:
on it!

nikitadhiman committed:
codestyle fix

nikitadhiman committed:
language string

nikitadhiman committed:
encountering invalid articleId using exception

This does not have any error code attached

nikitadhiman committed:
minor

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -733,7 +733,7 @@ public function shareToken($articleId)
 		if ($articleId == 0)
 		{
-			$this->setError(JText::_('COM_CONTENT_INVALID_ARTICLEID'));
+			throw new Exception(JText::_('COM_CONTENT_INVALID_ARTICLEID'));
```
Exception is too general, use InvalidArgumentException instead.

nikitadhiman committed:
exception

roland-d commented on administrator/components/com_content/models/article.php:
```diff
@@ -733,7 +733,7 @@ public function shareToken($articleId)
 		// checking for valid articleId
 		if ($articleId == 0)
```
Change this to a strict type check by doing this:

```
if ((int) $articleId === 0)
```

nikitadhiman committed:
limiting articleId check to int

nikitadhiman committed:
minor travis fix

zero-24 committed:
codestyle

zero-24 committed:
missing space

nikitadhiman committed:
comment update

zero-24 committed:
the comment

zero-24 committed:
the comment

nikitadhiman committed:
cs fix

codestyle for the branch share_token_generate
